### ISSUE TITLE
Command buffers, queues, queue families

### ISSUE THREAD

**Issue by Crisspl** [Devsh-Graphics-Programming/Nabla#3]
First of all: cmd buffers, queues and families are purely GPU objects, so I see no point in having CPU and GPU counterparts.
### Command pools
ICommandBuffer design is in the UML, however command buffers comes along with command pools. Command pools are opaque Vulkan objects used to minimize cost of acquaring/allocating resources for a set of cmd buffers assigned to a cmd pool. Also they in case of concurrency they require external synchronization. So what I'm thinking is that we don't expose command pools for users at all. Instead [under Vulkan backend], under the hood there can be created one command pool per queue family per thread. This can be done dynamically (when there's a need for such, i.e. cmd buffer is created by some thread using some queue family). This obviously means that all command buffers created by thread X and using queue family at index Y are "bounded to" the same command pool. In GL backend command pools would be non-existent at all.

### Major engine API change needed (maybe not that major)
I think we'll need to redefine what `device` is in Nabla. Currently is the main class to which (upon creation) we're passing what backend we want to use, then it's creating IVideoDriver, IAssetManager, IFileSystem etc... AFAIK Nabla doesnt let user choose the device, not even talking about using more than one device (GPU) at a time. There should be such possibility and then by `device` we mean something that will hold `VkPhysicalDevice` handle. `device` will create `driver` object that will hold `VkDevice` handle and, while creating, deciding what `queue` objects will be created (and more). 
* https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkDeviceQueueCreateInfo.html
* https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkDeviceCreateInfo.html

### Queue families
Same as Vulkan API does it. We can even use the same struct `VkQueueFamilyProperties`. We can query all queue families offered by current physical device with `vkGetPhysicalDeviceQueueFamilyProperties`. So at engine startup we query those to `smart_refctd_dynamic_array` or something and let user `device->getQueueFamilyProperties()` them. There's no need for anything like IQueueFamily engine-API objects.
In result API normalization between backends boils down to what is returned from `getQueueFamilyProperties()`. For Vulkan it's obvious. For GL we can for example report that there exists just one queue family with all (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkQueueFlagBits.html) capabilities with one queue. I dont know. Is there a way to query such things via GL API? Like whether our device supports graphics (maybe it's compute-only)? Not even asking about specific queues and their capabilities, I'm sure OpenGL hides it from user.

### Queues
If queue family properties are queried from device as described above (via `get*()`), then obviously `driver` creation must be deferred (not instantly created on `device` construction). IQueue objects are then created automatically by engine based on info given for `driver` creation (`VkDeviceQueueCreateInfo`s). Driver should have an API function to access them.

Actual UMLs and more design I'll post tomorrow.

**Comment by Crisspl** [comment]
https://app.genmymodel.com/api/repository/crisserpl2/Cmd%20bufs%20and%20queues

Things to watch out for: a single logical device (`driver`) can be created from multiple physical devices if all of them belong to the same device group.

**Comment by devshgraphicsprogramming** [comment]
> First of all: cmd buffers, queues and families are purely GPU objects, so I see no point in having CPU and GPU counterparts.

Maybe command buffers at some point but not right now.

**Comment by devshgraphicsprogramming** [comment]
> Major engine API change needed (maybe not that major)

That's another stage, see my code snipped about swapchains. Swapchains and devices need to be implemented together, right now we can treat `IDriver` or `IVideoDriver` as a default device.

> There's no need for anything like IQueueFamily engine-API objects.

`IQueueFamily` no, but `IQueue` yes.
We actually need the queues explicitly, even for the GL implementation, that's where the GL (offscreen) context will live.

>  For GL we can for example report that there exists just one queue family with all (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkQueueFlagBits.html) capabilities with one queue. I dont know. Is there a way to query such things via GL API? Like whether our device supports graphics (maybe it's compute-only)? Not even asking about specific queues and their capabilities, I'm sure OpenGL hides it from user.

Yes, OpenGL will only have one queue family.

**Comment by devshgraphicsprogramming** [comment]
> If queue family properties are queried from device as described above (via get*()), then obviously driver creation must be deferred (not instantly created on device construction). IQueue objects are then created automatically by engine based on info given for driver creation (VkDeviceQueueCreateInfos). Driver should have an API function to access them.

Why can't we first create the driver and then the queues?

**Comment by Crisspl** [comment]
> Why can't we first create the driver and then the queues?

Because of Vulkan API here 
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCreateDevice.html
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkDeviceCreateInfo.html
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkDeviceQueueCreateInfo.html
Queues we want to use have to be known at device creation. If you mean to create all possible queues as Vulkan objects, and then engine objects later - ofc it's possible. But I guess there is some advantage of not creating all of them if Vulkan gives such an option

As of API design, if above would be a good idea:
> IQueueFamily no, but IQueue yes.

Looking on your code snippet here https://github.com/buildaworldnet/IrrlichtBAW/issues/7#issuecomment-679280914 ...
```
auto* family = gpuAPI->getQueueFamily();
auto queue = family->createQueue();
```
This looks neat, so I think we could have `IQueueFamily` which would be just a container for `IQueue`s + family properties (flags, queue count...). So we have 2 options: this or not having `IQueueFamily` at all and operating on family indices and `SQueueFamilyProperties` only.

**Comment by devshgraphicsprogramming** [comment]
Thanks for pointing out that the queues need to be known before device creation, it seems that we need to be able to differentiate between an Instance an a Device just like how Vulkan does it.

I'm fine with declaring up-front how many queues (and swapchain) we'll have, this is actually how we do it now with the `AuxGLContexts`, you could see the auxilary contexts as a "kind-of" queue.

Makes it easier to manage OpenGL, the query about families can return only one queue family and its nice to know how many GL contexts we should create up-front.

**Comment by devshgraphicsprogramming** [comment]
We need to discuss semaphores, events and barriers (does vulkan have more sync primitives?)

There's also this thing called a timeline semaphore, which we will not bother to emulate because you cant even emulate it in lower Vulkan versions.

**Comment by devshgraphicsprogramming** [comment]
The logical device (`vkDevice`, not `vkPhysicalDevice`) will create **Semaphores, Fences and Events**

Naturally on OpenGL all need to be emulated with a `GLsync` object,

# Semaphores

Created by the logical device, starts in signallable state, **we will not allow external imported semaphores yet, and certainly not in OpenGL.**
The OpenGL signallable state means that the underlying `GLsync` object is missing/invalid, waitable means its valid, when semaphore is deleted and is in the waitable state, `glDeleteSync` must be called. In `DEBUG` mode if the semaphore is in the waitable state, and user is attempting to delete, we should call `glClientWaitSync` with a 0 timeout and assert that the fence is complete.


When a semaphore is signalled, a `glDeleteSync` is issued to release the previous GL object (if any), semaphore is back in the signallable state, then a `glFenceSync` call is issued after the appropriate point in the GL command stream and semaphore is in the waitable state, followed by a `glFlush`.

When a semaphore is waited upon, it is waited upon by `glWaitSync(fence,0,GL_TIMEOUT_IGNORED)` but only if the GLsync was placed on a different queue to the one waiting (same OpenGL context waiting on itself makes no sense in the presence of a `glMemoryBarrier`). Finally an appropriate `glMemoryBarrier` is always issued right after.

Even though `glWaitSync` will timeout after `GL_MAX_SERVER_WAIT_TIMEOUT` nanoseconds, that value is between 2^30 and 2^31 on all platforms that report it, so its probably safe to use it (to timeout, the time between GPU driver encountering the submitted wait command and the completion of the previously flushed commands followed by a signal, would have to be 1 second+). Furthermore the subsequent `glMemoryBarrier` will ensure some execution dependencies on the GPU.
Should we encounter any problems, we can always resort to a `glClientWaitSync` on a loop with a few us timeout before submitting commands to the GPU.


Queue Command Buffer Submissions to queues and sparse bindings take lists of semaphores to wait on and semaphores to signal.
Under OpenGL we take a bitwise OR of all `pWaitDstStageMask`, we need a function which maps destination `pair<VkPipelineStageFlags,VkAccessFlags>` (for QueueSumbit the accesss scope is everything) and to `GL_*_BARRIER_BIT` for use with `glMemoryBarrier`.

Queue Swapchain Presents takes a list of semaphores to wait on.

Swapchain Acquire takes only one semaphore to signal,

**Comment by devshgraphicsprogramming** [comment]
# Fences

**Comment by devshgraphicsprogramming** [comment]
# Events

**Comment by devshgraphicsprogramming** [comment]
# Pipeline Barriers ( and FUTURE Subpass Dependencies )

A pipeline barrier is basically a `glMemoryBarrier` with appropriate barrier bits deduced from the arguments to `vkCmdPipelineBarrier`. There will always ever be 1 `glMemoryBarrier` because OpenGL does not do that granularity level tracking.

We can issue `glMemoryBarrierByRegion` instead of `glMemoryBarrier` here, but we actually dont need to issue `glMemoryBarrier` that frequently in the first place. Also `*ByRegion` is only available under GL 4.5, ES 3.1 or with `ARB_ES3_1_compatibility`.

Basically we need a painfully complicated function that given the source and destination stage masks + memory barriers (buffer, image, etc) access masks, can figure out the bitfields on `glMemoryBarrier` to use.

**Comment by devshgraphicsprogramming** [comment]
@Crisspl PR opened against old repo 
https://github.com/buildaworldnet/IrrlichtBAW/pull/540

**Comment by devshgraphicsprogramming** [comment]
We should have a separete `COpenGLSync` to allow for refcounted OpenGL sync objects, so that there's no hundreds of `glFenceSync` being placed when there are hundreds of semaphores, fences and events to signal.

**Comment by devshgraphicsprogramming** [comment]
Closing but we should probably open another issue for emulating Events and figuring out the PipelineBarriers which are local (by region/viewport)
